src/server.h

```diff
@@ -2163,6 +2163,9 @@ int ACLAuthenticateUser(client *c, robj *username, robj *password);
 unsigned long ACLGetCommandID(const char *cmdname);
 void ACLClearCommandID(void);
 user *ACLGetUserByName(const char *name, size_t namelen);
+int ACLCheckCommand(client *c, struct redisCommand *cmd, const char *subcmd);
+int ACLCheckKey(client *c, const char *key, int keylen);
+int ACLCheckPubsubChannelPerm(sds channel, list *allowed, int literal);
```

Comment:
Do you need to have such low level access to determine what the user does and does not have access to. I would expect it to be easier to maintain to just have an ACLCheckClientPermission(client, command arguments), so that it emulates the normal permission check that we do.

Reply:
@madolson a module may want to check permission to a key before doing RM_OpenKey (so the low level CheckKey api is needed).
in other cases before executing a sequence of RM_Calls, the module will want to first make sure they're all allowed.

i haven't yet looked at the code, but while responding to this message i see the signature of the CheckCommand interface, and i think it's insufficient, we want to pass the full argv array, so keys will get checked too (in some cases the module may be a "script engine" and it's doesn't really understand what command it executes and which arguments refer to key names).

Reply:
@oranagra Maybe my questions are dumb and this is obvious to you module developers, but I don't follow why having three individual commands is useful.

> a module may want to check permission to a key before doing RM_OpenKey (so the low level CheckKey api is needed).

I don't understand this use case, in what context would this be used instead of having the module declare what keys are needed by the command? A lot of code in Redis (cluster mode and server-side key tracking) assume this. It seems weird to have a separate backchannel for defining if the user has access. It also seems weird that this won't be published to the ACL log.

> in other cases before executing a sequence of RM_Calls, the module will want to first make sure they're all allowed.

Not sure I understand why you would want three separate API calls here instead of just one that answers the question, "Can this user execute this command". It's fragile to future changes where we add more ACL changes.

Reply:
As we know, by design, modules are exempt from ACL validations on the actions they do.
i.e. Antirez said that since they can write to arbitrary memory and execute arbitrary syscalls, when a user allowed a certain module command to run, it can do whatever it wants with no limitations from redis, and this includes doing any RM_Call they like.

So the first obvious thing is to add the flag to RM_Call that enables full checks inside it.
but as i stated, in some cases the module wants to verify that a sequence of commands can succeed before issuing the first one, so that's why we need the first new API.

As for the other ones, some modules access keys that are not explicitly referred to in the argument list.
i can provide few trivial cases:
1. a module that implements a domain specific query language that has some operators before each argument or isn't at all separated by spaces or distinct arguments
2. a module that for some reason adds prefixes to keys names it got from arguments, or some other funny shit (think of SORT BY).
3. a module that implements a scripting language, for instance one that exposes similar interface to what EVAL has today, so the user passes a script, and maybe also a list of key names to be verified before the module command is executed, but there's no guarantee that the script doesn't access other keys.

needless to say, some of the examples above are not compatible with cluster mode.

So i'm aiming to do two things:
1. provide something that's easy to use and hard to mis-use.
2. provide low level tools for advanced modules.

@yossigo @MeirShpilraien please chime in.
@YaacovHazan please update RM_ACLCheckCommandPerm to take a full argv list and validate all that's validated by processCommand

Reply:
1. I suppose we should add a low level cluster mode command to check for cross slots then as well.
2. I think what you were expecting with RM_ACLCheckCommandPerm was the alternative I was asking about originally. Having a module check that does the EXACT permission check that Redis normally seems easier to use and maintain. I'm fine to also have other APIs for power users to do shenanigans with.